File: js/fraction.js
class Fraction extends RationalNumber {
    constructor(numerator, denominator) {
        super();
        this.object = new QuotientObject(new IntegerNumber(numerator), new IntegerNumber(denominator));
    }

    get() {
        return this.numerator() / this.denominator();
    }

    numerator() {
        return this.object.dividend.get();
    }
    denominator() {
        return this.object.dividor.get();
    }

    static get(n) {
        if(!RationalNumber.is(n)) return;

        let numerator = n;
        let denominator = 1;
        
        do {
            numerator *= 10;
            denominator *= 10;
        } while(!IntegerNumber.is(numerator));

        return Fraction.reduce(new Fraction(numerator, denominator));
    }

    static parse(n) {
        if(n.constructor.name == "IntegerNumber") return new Fraction(n.get(), 1);
        if(n.constructor.name == "Fraction") return n;
    }

    static sum(f1, f2) {
        let numerator = f1.numerator() * f2.denominator() + f2.numerator() * f1.denominator();
        let denominator = f1.denominator() * f2.denominator();
        return Fraction.reduce(new Fraction(numerator, denominator))
    }

    static difference(f1, f2) {
        return Fraction.sum(f1, Fraction.parse(_Number.additiveInverse(f2)));
    }

    static multiply(f1, f2) {
        return Fraction.reduce(new Fraction(f1.numerator() * f2.numerator(), f1.denominator() * f2.denominator()));
    }

    static divide(f1, f2) {
        return Fraction.multiply(f1, Fraction.parse(_Number.multiplicativeInverse(f2)));
    }

    static reduce(n) {
        let gcd = _Number.GCD(n.numerator(), n.denominator());
        
        let numerator = n.numerator() / gcd;
        let denominator = n.denominator() / gcd;
        
        return new Fraction(numerator, denominator);
    }
}

File: cpp/Fraction.cpp
#include <string>

#include "Fraction.h"
#include "Integer_Number.h"

Fraction::Fraction(double n) : Fraction(Fraction::get(n)) {
}
Fraction::Fraction(Rational_Number n) : Fraction(n.get_numerator(), n.get_denominator()) {
}

double Fraction::get() {
	return Number::get();
}

Fraction Fraction::reduce(Fraction f) {
	return (Fraction)Rational_Number::reduce(f);
}

Fraction Fraction::common_denominator(Fraction f1, Fraction f2) {
	return (Fraction)Rational_Number::common_denominator(f1, f2);
}

Fraction Fraction::get(double n) {
	double numerator = n;
	double denominator = 1;

	do {
		numerator *= 10;
		denominator *= 10;
	} while (!Integer_Number::is(numerator));

	if (denominator < 0) {
		numerator *= -1;
		denominator *= -1;
	}
	return Fraction::reduce(Fraction((int)numerator, (unsigned int)denominator));
}

std::ostream& operator<<(std::ostream& out, Fraction f) {
	out << f.decimal_form();
	return out;
}

std::string Fraction::simple_form() {
	if (Integer_Number::is(n)) return std::to_string((int)n);
	else return std::to_string(get_numerator()) + "/" + std::to_string(get_denominator());
}
std::string Fraction::decimal_form() {
	return to_string();
}